### ISSUE TITLE
Proposal: Generating ./geometry/*.root during build

### ISSUE THREAD

**Issue by klenze** [R3BRootGroup/R3BRoot#335]
From what I understand, the geometry root files are generated from macros/geo/create_${DETECTOR}.C. 

Still, they are added to the repository. This could cause some possible problems:
* Incremental changes become opaque. From gits point of view, a binary blob is just replaced by another binary blob. 
* There is always the exciting possibility that the generating macros and the binary root geometry file diverge, because someone added the binary change in the main repository but not the macro change in the macros repository, or vice versa.  
* Furthermore, at least the floating point operations used to generate the geometries are [implementation dependent](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html#3098). So "let's run the create script and do a diff" is right out. One would have to do a semantic comparision. 
* Finally, this seems to encourage duplication. (See #317 .) At the moment, we have 14 geometries for califa, and 9 different create_califa_*.C scripts. Given that I had to re-add support for the 2019/s444 geometry, I would wager that most versions will fail to work with the current code base. 

The largest file in geometries is some 150kBytes. I think it is both feasible and desirable to auto-generate the supported versions of these files (e.g. 2019, 2020) during the cmake build process. 

What do you think?

**Comment by hapol** [comment]
To my understanding the files in macros/geo/  are under the code control and evolve with time, while the existing root files should not be modified, as they correspond to particular situations, and therefore they do not evolve. If a root file is modified, then it is stored under other name. If this cause a problem (space, info, ...), it could be modified.
So, regarding the problems:
- There should be no incremental changes in the root files, they represent some stable geometries.
- It's true that they can diverge, and it would be great to have a kind of meta-file or README specifying how each useful root file is created from a (git-labelled) macro.
- Is this true? is there any case where you have seen a difference in a root file?
- As you know, the new CALIFA code is only supporting CALIFA_2020 geometry. Other legacy geometries are still not supported or require an older R3BRoot version. So, it would be necessary to clean...

**Comment by janmayer** [comment]
I see this issue as well, and would like to raise some further points:

> There should be no incremental changes in the root files, they represent some stable geometries.
- I agree that they do represent a stable geometry, but they are produced from a macro - thus that macro they were created from is the actual stable configuration. If this is macro, or the way in which it was invoked, is lost (which might have already happened), it is hard to reproduce the root file or analyze what it entails / what changed.

> There is always the exciting possibility that the generating macros and the binary root geometry file diverge, because someone added the binary change in the main repository but not the macro change in the macros repository, or vice versa.
> Other legacy geometries are still not supported or require an older R3BRoot version. So, it would be necessary to clean...
- In addition, the detector class might develop as well and not work with the older geometries - all combinations should be tested (wishful thinking).

>  I think it is both feasible and desirable to auto-generate the supported versions of these files (e.g. 2019, 2020) during the cmake build process.
- I would go even further and propose that the detector simulation classes could create the geometry on-demand. For example, with NeuLAND, I already have an option to ask the Simulation class to [get the right geometry for a number of double planes and move it into position](https://github.com/R3BRootGroup/R3BRoot/blob/dev/neuland/simulation/R3BNeuland.cxx#L72) without having to worry about the file name or version. It would not be a large step to have it produce the geometry from scratch on-the-fly.
- I think it is semantically not correct to have the geometry-creating-macros under `/marcos/geo`. I would say that they (or most of them) belong closely to the individual detector and probably should be under `/<detector>/geometry` oder `/<detector>/simulation`. *


*) Tangent topic: As there are so many folders in the root dir, it has become quite cluttered. In general, I would propose moving all detectors to `/detectors/` or at least put all the fibers under `/fibers/<fiXX>`. What is the difference between all those fiber classes anyway? Would a template help?

**Comment by vadimr3b** [comment]
I would agree with Jan and produces the geometries on the fly for the simulation.
If we have different states for the detector, we should rather solve this by git tags (e.g. s444).
Like this you can do a simulation with the setup matching your experiment by simply checking out the tag and everything should work.

I also agree to Jans * ;)

**Comment by janmayer** [comment]
> If we have different states for the detector, we should rather solve this by git branchs (e.g. s444).
Or/And individual repositories for each experiment instead of that big macro repo.

**Comment by klenze** [comment]
The idea for this issue came to me when I read 
(this code)[https://github.com/R3BRootGroup/R3BRoot/blob/c578178be870aba7de10697c339598e7056b10aa/califa/R3BCalifaGeometry.cxx#L255]

```
int R3BCalifaGeometry::GetCrystalId(const char* volumePath)
{
[...]
    Int_t alveolusCopy = atoi(volumePath + 34); 
    // converting to int the alveolus copy
    if (alveolusCopy < 10)
        cryType = atoi(volumePath + 76); // converting to int the crystal type
    else
        cryType = atoi(volumePath + 77); // converting to int the crystal type
```

What happens here is that we try to extract numbers from a string in a geometry file. As the string was created using the format specifier %d in one place instead of e.g. %02d, we now have to work around this and catch the case where the first number is larger than 10. 

The lamentable need to do string parsing in the inner loop of the simulation non-withstanding, a better solution would have been to update the califa 2020 geometry to use %02d. If geometry files were though of as easily generated, transient files (such as the cxx.o files are), this update would have been easy. If they are instead thought of as fixed, essential components of our project, we will always need workarounds like that. 

I would furthermore propose to move both the generation macros (and the macros used in test cases) to the main repository, so that there is a clear correspondence between macro versions and detector code versions. This would take a lot of detective work (e.g. "this main repo commit was followed by that macros commit by the same author, so the two versions are probably intended to work with each other") out of the job of people wanting to run old versions. (We can still keep the macros subrepository as a Schmuddelecke for code which is neither compiled nor tested against the latest version of the detector classes and thus will probably not work.)

While talking about hypothetical changes, I would also propose making "dev" the default branch of R3BRoot, imposing a function length limit of 30 lines, reject any PRs containing the string "sprintf" with prejudice and providing all code authors with fusion powered jetpacks. :-)

**Comment by hapol** [comment]
I guess you are at the wrong forum making your suggestion: your comment "using the format specifier %d in one place instead of e.g. %02d" and "The lamentable need to do string parsing in the inner loop of the simulation non-withstanding, a better solution would have been to update the califa 2020 geometry to use %02d" correspond to features of the TGeoVolume::AddNode() function in ROOT VMC, and not to the macros that we made for the CALIFA geometry. https://root.cern.ch/doc/master/TGeoVolume_8cxx_source.html#l00930

As the only way to retrieve the object is using the volume path, we should perform the string parsing. As I told you previously, if you know another method to access to the volume information, please, let us know.

**Comment by janmayer** [comment]
Yeah, the ROOT geometry stuff is extremely bad - and to think this terrible interface is at the core of our simulations... Can we just drop ROOT Geo and just use pure Geant4?

On Topic: Could you use `gMC->CurrentVolOffID(1, fPaddleID);` like [here](https://github.com/R3BRootGroup/R3BRoot/blob/dev/neuland/simulation/R3BNeuland.cxx#L114)?

**Comment by klenze** [comment]
> I guess you are at the wrong forum making your suggestion: your comment "using the format specifier %d in one place instead of e.g. %02d" and "The lamentable need to do string parsing in the inner loop of the simulation non-withstanding, a better solution would have been to update the califa 2020 geometry to use %02d" correspond to features of the TGeoVolume::AddNode() function in ROOT VMC, and not to the macros that we made for the CALIFA geometry. https://root.cern.ch/doc/master/TGeoVolume_8cxx_source.html#l00930

Ooops. I stand corrected. Sorry.
> 
> As the only way to retrieve the object is using the volume path, we should perform the string parsing. As I told you previously, if you know another method to access to the volume information, please, let us know.

I do not have a method which does not break the TVirtualMC abstraction. :-(
(Creating a new class extending  TGeoVolume to include a CrystalID field might be an option, but I would not count on TGeoWhatever supporting this).

That being said, I noticed a speedup by several orders of magnitudes when I stopped simulating 4.4GeV photons and started with 4.4MeV photons instead, so my complaints about speed are mostly a thing of the past now. :-)

**Comment by inkdot7** [comment]
> > If we have different states for the detector, we should rather solve this by git branchs (e.g. s444).
> 
> Or/And individual repositories for each experiment instead of that big macro repo.

General remark: branches are useful for code development, not history tracking of a setup.

Please do not use git branches to try to handle things that code-wise shall survive for a long time.  If there were two different geometries in use at different experiments, both (generating macros!) should be in the code, in (all later) branches.  (Even better: not separate macros, but options within the macro to handle the differences.)

If needed, then separate repositories for different experiments.  (With only the experiment-specific stuff of course, not duplicating/forking the big project.)

**Comment by klenze** [comment]
> Yeah, the ROOT geometry stuff is extremely bad - and to think this terrible interface is at the core of our simulations... Can we just drop ROOT Geo and just use pure Geant4?

This, 100%. 

Today, while waiting for a 10M event simulation to finish, I tried to get rid of the ">>> Event " ... lines our simulation includes once per event. While there is obviously a simple way to do that which I am totally missing, I had to:

0.) break abstraction by casting TVirtualMC to TGeant4
1.) break access specifiers to get to TGeant4::fRunManager
2.) break access specifiers to TG4RunManager::fRunManager, which is the actual G4RunManager
3.) dynamic_cast the result of G4RunManager::GetUserEventAction() to const TG4EventAction*
4.) const_cast that to TG4EventAction*
5.) break access specifiers to call TG4EventAction::VerboseLevel()

So, all in all, it just took five evil things (excluding step 3) to get rid of the spam. Needless to say, by the time I was done, so was my simulation. 

Also notice that @janmayer, who proposed moving to raw Geant4 is a NeuLAND person. From my point of view, this means that the argument "We need to support Geant3 because it handles neutrons better" is probably no longer true.

`\begin{soapbox}`
In general, my attitude towards using 'the ROOT way of doing something' is rather dissimilar to that of the ${EXP}Root projects. Personally, I would crawl through miles of desert to get from TList to std::list, TVirtualMC to Geant4, TFormula to lambda expressions and so on. I can pragmatically accept that for some tasks (plotting histograms, handling event data), ROOT is a viable solution, but I am not really in the "ROOT macros can do anything, so let's use them to store our simulation configuration" camp. 
`\end{soapbox}`

**Comment by janmayer** [comment]
> Also notice that @janmayer, who proposed moving to raw Geant4 is a NeuLAND person. From my point of view, this means that the argument "We need to support Geant3 because it handles neutrons better" is probably no longer true.

Thank you for the confidence in my expertise, but that's maybe a bit too much.

I simply don't use Geant3 anymore - its super annoying to install, as the "modern" fortran compilers don't want to compile it anymore, i.e. I have to change my devtoolset during the compilation of FairSoft. It often simply refuses to work for me.

It does give *different* results from Geant4, depending on the physics lists, but I can't really tell if they are *better* - but I can tell that Geant4 is updated and benchmarked regularly by a large user base.

Overall I simply think Geant3 is simply obsolete, and with it so is TVirtualMC and the whole ROOT geometry module.

**Comment by inkdot7** [comment]
@klenze - some text to describe to the reader about when / with which commits this issue was resolved?

**Comment by klenze** [comment]
@inkdot7: Sorry, I was just cleaning up my stuff. 

In the meantime, @jose-luis-rs has implemented automatic generation of ``${CMAKE_SOURCE_DIR}/geometry/`` geometry root files during the cmake invocation. 

At least for CALIFA, the correctness of the result very much seems to depend on the exact platform used. So it seems that without a way to check if a geometry file is good, auto-creating geometry files might not be a good idea. 

If you have an idea of how to check if two compressed, opaque root files contain _compatible_ floating point values, that would be helpful. Personally, I find the [syntax](https://github.com/R3BRootGroup/R3BRoot/blob/dev/califa/pars/R3BCalifaGeometry.cxx#L150) to extract even the center of a crystal a bit byzantine compared to [plain geant4](https://apc.u-paris.fr/~franco/g4doxy/html/classG4VPhysicalVolume.html#21e40e5435cc6a506207d77f550e82fb).

Jan's suggestion to move to plain Geant4 would be my preferred solution as well, but this is very much a minority view.

**Comment by inkdot7** [comment]
> At least for CALIFA, the correctness of the result very much seems to depend on the exact platform used.

Does this mean that actual results (e.g. simulated spectra) vary a lot (beyond random fluctuations) (= bug), or 'just' that the generated geometry files have slightly different values due to rounding differences?

I do not have anything to compare generic .root files for compatibility, but if spectra can be generated, this fuzzy spectra CI helper may be of some use:

https://git.chalmers.se/expsubphys/fuzzyspectra

**Comment by jose-luis-rs** [comment]
The ROOT geometries are generated dynamically during compilation and the generation process seems to work correctly.
We still need to include a CI test to ensure that the results are always reproducible, but this is currently on the TODO list.
If nobody volunteers to take care of this implementation, I will handle it after my teaching break.

>> Today, while waiting for a 10M event simulation to finish, I tried to get rid of the ">>> Event " ... lines our simulation includes once per event. While there is obviously a simple way to do that which I am totally missing, 

Here you can find the Geant4 commands to do it in a simple way https://github.com/R3BRootGroup/R3BRoot/pull/1244

I close this issue, reopen it if needed.

**Comment by YanzhaoW** [comment]
As I have presented in the collaboration meeting, neuland simulation is now building its geometry during its run-time without generating the file. This gives much better flexibility to the simulation process.

> The ROOT geometries are generated dynamically during compilation

If not wrong, I think they are generated during the cmake configuration?

There is one pitfall about this: Users have to setup the environment variables correctly. This means the root executable must exist in `PATH` and FairSoft and FairRoot shared library folders must also exist in `LD_LIBRARY_PATH`.

**Comment by jose-luis-rs** [comment]
Yes, you are right. The configuration file must be loaded before running CMake. For GSI users, this means the usual conf.sh file located outside the build folder.

> As I have presented in the collaboration meeting, NeuLAND simulation is now building its geometry during its run-time without generating the file. This gives much better flexibility to the simulation process.

I think this is a new implementation and that it’s not in the dev branch, right?
At the moment, we need three geometries for the current simulations, with 12, 13, and 30 DPs, and the first DP must always be in the same position for all geometries. Of course, I assume that in your new implementation the user will be able to generate all possible configurations for NeuLAND, so all of them should fulfill that condition.

**Comment by YanzhaoW** [comment]
Hi, @jose-luis-rs 

> I think this is a new implementation and that it’s not in the dev branch, right?

No, it's in the feature branch #1201 . The change is actually very subtle:

Before, we created the geometry nodes by a function in a MACRO and then save them into a root file. When launching the simulation,  the overriden virtual function of `FairDetector::ConstructGeometry` will read the root file and copy the nodes to the `gGeomanager`.

For the new implementation, we change the implementation of  `FairDetector::ConstructGeometry` such that nodes are directly created (using the same function) in `gGeomanager` when launching the simulation.

> I assume that in your new implementation the user will be able to generate all possible configurations for NeuLAND, so all of them should fulfill that condition.

Yes, users just need to give any number of double planes.
